File: L17AdobeMedium/Triangle.java

import java.util.ArrayList;
import java.util.List;

public class Triangle {

    public int minimumTotal(List<List<Integer>> triangle) {
        int n = triangle.size();
        // Start with the last row in the triangle
        List<Integer> dp = new ArrayList<>(triangle.get(n - 1));

        // Traverse from the second-last row to the top
        for (int i = n - 2; i >= 0; i--) {
            for (int j = 0; j < triangle.get(i).size(); j++) {
                // Update dp[j] with the minimum path sum from below
                dp.set(j, triangle.get(i).get(j) + Math.min(dp.get(j), dp.get(j + 1)));
            }
        }

        // The top element of dp will hold the minimum total
        return dp.get(0);
    }

    public static void main(String[] args) {
        Triangle solution = new Triangle();

        // Initialize the triangle as a list of lists
        List<List<Integer>> triangle = new ArrayList<>();
        triangle.add(List.of(2));
        triangle.add(List.of(3, 4));
        triangle.add(List.of(6, 5, 1));
        triangle.add(List.of(4, 1, 8, 3));

        // Get the minimum total path sum
        int minTotal = solution.minimumTotal(triangle);
        System.out.println("Minimum Total: " + minTotal);
    }
}

File: L17AdobeMedium/Triangle.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    int minimumTotal(vector<vector<int> > &triangle)
    {
        int n = triangle.size();
        vector<int> dp(triangle[n - 1]); // Start with last row

        for (int row = n - 2; row >= 0; row--)
        { // Bottom-up approach
            for (int col = 0; col < triangle[row].size(); col++)
            {
                dp[col] = triangle[row][col] + min(dp[col], dp[col + 1]);
            }
        }

        return dp[0]; // The minimum path sum
    }
};

int main()
{
    Solution solution;

    vector<vector<int> > triangle = {
        {2},
        {3, 4},
        {6, 5, 1},
        {4, 1, 8, 3}};

    int minTotal = solution.minimumTotal(triangle);
    cout << "Minimum Total: " << minTotal << endl;

    return 0;
}


//input parameters
//     2
//    3 4
//   6 5 1
//  4 1 8 3



// Starting from the bottom row:
// Row 3 (Base): [4, 1, 8, 3]
// Row 2: [6+1=7, 5+1=6, 1+3=4] → [7, 6, 4]
// Row 1: [3+6=9, 4+6=10] → [9, 10]
// Row 0: [2+9=11] → [11]




// Time & Space Complexity
// Time Complexity: O(n²) (Iterates through the triangle)
// Space Complexity: O(n) (Optimized to a single array)
